### PR TITLE
HPCC-23975 Convert act. stats to CThorStatsCollection

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jutil.hpp [hunk 1]
```diff
@@ -565,5 +566,17 @@ inline bool stdContains(Container&& container, Value &&v)
...
+class jlib_decl COnScopeExit
+{
+    const std::function<void()> &exitFunc;
+public:
+    inline COnScopeExit(const std::function<void()> &_exitFunc) : exitFunc(_exitFunc) { }
+    inline ~COnScopeExit()
+    {
+        exitFunc();
+    }
+};
```
Comment (jakesmith) [inline system/jlib/jutil.hpp:579]: @ghalliday - I couldn't see an equivalent of this, it's a frequent pattern, recoded in various places.
Could be a separate PR ?

**Comment by jakesmith** [comment]
@ghalliday - please review

File: system/jlib/jutil.hpp (hunk 1 above)
Comment (ghalliday) [reply system/jlib/jutil.hpp:579]: Looks like a useful addition.

**Review by ghalliday — COMMENTED** [review]
@jakesmith I'm just wondering about this.  In roxie the activities have a single set of stats defined in the base class with the mapping passed from the derived activity constructor.  I think that would probably make the code simpler (avoiding derived classes overriding deserialize etc.).

**Comment by jakesmith** [comment]
>I'm just wondering about this. In roxie the activities have a single set of stats defined in the base class with the mapping passed from the derived activity constructor. I think that would probably make the code simpler (avoiding derived classes overriding deserialize etc.).

I also wondered about that.
I was a bigger change though and decided to go this route for now.

Perhaps assuming not other issues, will refactor to consolidate further in a separate PR?
@ghalliday

File: thorlcr/activities/indexread/thindexreadslave.cpp [hunk 2]
```diff
@@ -80,6 +77,24 @@ class CIndexReadSlaveBase : public CSlaveActivity
...
+        unsigned __int64 startSeeks = 0, startScans = 0;
+    public:
+        inline CCaptureIndexStats(CRuntimeStatisticCollection &_stats, StatProvider &_statProvider) : stats(_stats), statProvider(_statProvider)
+        {
+            startSeeks = statProvider.querySeeks();
+            startScans = statProvider.queryScans();
+        }
+        inline ~CCaptureIndexStats()
+        {
+            stats.mergeStatistic(StNumIndexSeeks, statProvider.querySeeks() - startSeeks);
```
Comment (ghalliday) [inline thorlcr/activities/indexread/thindexreadslave.cpp:94]: Slightly concerned about the extra overhead of this call.  Might be worth considering a fastpath for this particular case of summing a stat.

File: thorlcr/activities/keyedjoin/thkeyedjoin.cpp [hunk 3]
```diff
@@ -37,7 +36,9 @@ class CKeyedJoinMaster : public CMasterActivity
     MemoryBuffer initMb;
     unsigned numTags = 0;
     std::vector<mptag_t> tags;
-    ProgressInfoArray progressInfoArr;
+    unsigned numStats = 0;
```
Comment (ghalliday) [inline thorlcr/activities/keyedjoin/thkeyedjoin.cpp:39]: I don't think this is actually used.

**Review by ghalliday — COMMENTED** [review]
@jakesmith I've scanned it.  It looks like it would be quite a bit simpler if you passed the mapping to the base, but I can review fully if you want to merge as is.

**Comment by jakesmith** [comment]
>It looks like it would be quite a bit simpler if you passed the mapping to the base, but I can review fully if you want to merge as is.

I'll take a look first, perhaps I just had refactor fatigue.

**Comment by ghalliday** [comment]
BTW Richard merged https://github.com/hpcc-systems/HPCC-Platform/pull/13712 which could be used in a couple of places in your code.

**Comment by jakesmith** [comment]
@ghalliday - please review. NB: left 1st commit as it was. 2nd commit is further refactoring to amalgamate the stats into a single collection container in the base, with a single serialize/deserialize.

File: system/jlib/jstats.cpp [hunk 4]
```diff
@@ -1236,7 +1206,18 @@ void StatisticsMapping::createMappings()
...
+{
+public:
+    AllStatisticsMapping()
+    {
+        for (int i = StKindAll+1; i < StMax; i++)
+            indexToKind.append(i);
+        createMappings();
+    }
+} allStatisticsMapping;
+const StatisticsMapping &allStatistics = allStatisticsMapping;
```
Comment (ghalliday) [inline system/jlib/jstats.cpp:1220]: What is the reason for this change?

File: system/jlib/jstats.cpp [hunk 5]
```diff
@@ -1236,7 +1206,18 @@ void StatisticsMapping::createMappings()
     }
 }
 
-const StatisticsMapping allStatistics;
+
+class AllStatisticsMapping : public StatisticsMapping
```
Comment (ghalliday) [inline system/jlib/jstats.cpp:1210]: What is the advantage in having a different mapping class?

File: system/jlib/jstats.h [hunk 6]
```diff
@@ -400,13 +400,18 @@ class jlib_decl StatisticsFilter : public CInterface
...
-    //Takes a list of StatisticKind
-    StatisticsMapping(const std::initializer_list<StatisticKind> &kinds);
-    //Takes an existing Mapping, and extends it with a list of StatisticKind
-    StatisticsMapping(const StatisticsMapping * from, const std::initializer_list<StatisticKind> &kinds);
-    //Accepts all StatisticKind values
-    StatisticsMapping();
-
+    //Takes a list of StatisticKind and a variable number of existing mappings and combines
+    template <typename... Mappings>
+    StatisticsMapping(const std::initializer_list<StatisticKind> &kinds, const Mappings &... mappings) : StatisticsMapping(&mappings...)
```
Comment (ghalliday) [inline system/jlib/jstats.h:405]: Changing to a reference makes sense, but why reverse the order, and why inline the function (I guess because it is now a template class?)

File: system/jlib/jstats.cpp (hunk 4 above)
Comment (jakesmith) [reply system/jlib/jstats.cpp:1220]: It's related to the changes in the constructor of StatisticsMapping.
Before the default constructor was the 'all' types constructor.
With the variadic list on the constructor, the default constructor had to be nop.

File: system/jlib/jstats.h (hunk 6 above)
Comment (jakesmith) [reply system/jlib/jstats.h:405]: > why reverse the order

because the tail is a variadic list of existing StatisticMapping's.
e.g. for use case in thor like:
```
const StatisticsMapping joinActivityStatistics({StNumLeftRows, StNumRightRows}, basicActivityStatistics, spillStatistics);
```

>why inline the function (I guess because it is now a template class?)
yes has to be inline because it's a template.

**Comment by ghalliday** [comment]
@jakesmith first impression is that it simplifies the code significantly.  I'm not sure why you have made the all stats change and related items.  (See questions.)  I'll take a look at the commits properly tomorrow.

File: system/jlib/jstats.cpp (hunk 5 above)
Comment (ghalliday) [reply system/jlib/jstats.cpp:1210]: ok, I understand from the weird variadic constructor.

  Another alternative is to use a different constructor.  E.g., a single stat, which if it is StKindAll adds them all.

File: system/jlib/jstats.h (hunk 6 above)
Comment (ghalliday) [reply system/jlib/jstats.h:405]: Very clever, and painful to my brain.

File: system/jlib/jstats.cpp (hunk 5 above)
Comment (jakesmith) [reply system/jlib/jstats.cpp:1210]: yes considered an alternate constructor, but didn't think of a logical one to create.
Your suggestion would work though, if you prefer can change it to that.

**Comment by jakesmith** [comment]
@ghalliday - will tag you back for now. Let me know if you think it would be better to have a separate constructor for 'all' , vs AllStatisticsMapping.

File: system/jlib/jstats.cpp (hunk 5 above)
Comment (ghalliday) [reply system/jlib/jstats.cpp:1210]: I think I would prefer only having a single StatisticsMapping class, so yes please if you are willing.  I don't mind what form the constructor takes (within reason!).

File: thorlcr/activities/join/thjoinslave.cpp [hunk 7]
```diff
@@ -458,7 +457,10 @@ class JoinSlaveActivity : public CSlaveActivity, implements ILookAheadStopNotify
             leftStream.setown(iLoaderL->load(leftInputStream, abortSoon));
             isemptylhs = 0 == iLoaderL->numRows();
             stopLeftInput();
+
+            CRuntimeStatisticCollection spillStats(spillStatistics);
```
Comment (ghalliday) [inline thorlcr/activities/join/thjoinslave.cpp:461]: We should introduce
mergeStats(spillStats, iLoadedL, spillStatistics) to avoid the temporary StatsCollection.
Created HPCC-24004 to implement once this is merged

File: thorlcr/activities/keyedjoin/thkeyedjoinslave-legacy.cpp [hunk 8]
```diff
@@ -2418,11 +2418,43 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
...
+    virtual void serializeStats(MemoryBuffer &mb) override
     {
-        CSlaveActivity::serializeStats(mb);
         ForEachItemIn(s, _statsArr)
-            mb.append(_statsArr.item(s));
+        {
+            unsigned __int64 v = _statsArr.item(s);
+            if (0 == v)
+                continue;
+            switch (s)
```
Comment (ghalliday) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave-legacy.cpp:2428]: simplify: better with a
constexpr StatisticKind mapping[] = { StNumIndexSeeks, StNumIndexScans, ... };
and index into that to provide the mapping.
stats.setStatistic(mapping[s]);

**Review by ghalliday — COMMENTED** [review]
@jakesmith a few minor comments.  It looks like it has cleaned up the code.  I have created a jira HPCC-24004 to cover one further improvement.

**Comment by jakesmith** [comment]
@ghalliday - please see changes in last commit.
